internal/controller/nimpipeline_controller.go

```diff
@@ -33,30 +39,257 @@ type NIMPipelineReconciler struct {
+	// TODO: Add custom cleanup
+
+	// All owned objects are garbage collected
+	return nil
+}
+
+func (r *NIMPipelineReconciler) reconcileNIMPipeline(ctx context.Context, nimPipeline *appsv1alpha1.NIMPipeline) (ctrl.Result, error) {
+	// Track enabled NIMServices
+	enabledServices := make(map[string]bool)
+
+	// Process each service specification in the pipeline
+	for _, service := range nimPipeline.Spec.Services {
+		enabled := service.Enabled != nil && *service.Enabled
+		enabledServices[service.Name] = enabled
+
+		if !enabled {
+			continue
+		}
+
+		if err := r.reconcileNIMService(ctx, nimPipeline, service); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Clean up disabled NIMServices
+	if err := r.cleanupDisabledNIMs(ctx, nimPipeline, enabledServices); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Update status of NIMServices
+	if err := r.updateStatus(ctx, nimPipeline, enabledServices); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *NIMPipelineReconciler) injectDependencies(nimService *appsv1alpha1.NIMService, dependencies []appsv1alpha1.ServiceDependency) {
+	for _, dep := range dependencies {
+		var serviceEnvVars []corev1.EnvVar
+
+		// Check if custom environment variable names and values are provided and inject them
+		if dep.EnvName != "" && dep.EnvValue != "" {
+			// Use custom environment variables
+			serviceEnvVars = []corev1.EnvVar{
+				{
+					Name:  dep.EnvName,
+					Value: dep.EnvValue,
+				},
+			}
+			// Merge and inject the environment variables
+			nimService.Spec.Env = utils.MergeEnvVars(nimService.Spec.Env, serviceEnvVars)
+		}
+	}
+}
+
+func (r *NIMPipelineReconciler) reconcileNIMService(ctx context.Context, nimPipeline *appsv1alpha1.NIMPipeline, service appsv1alpha1.NIMServicePipelineSpec) error {
+	logger := log.FromContext(ctx)
+
+	nimService := &appsv1alpha1.NIMService{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      service.Name,
+			Namespace: nimPipeline.Namespace,
+		},
+		Spec: service.Spec,
+	}
+
+	// Inject service dependencies
+	r.injectDependencies(nimService, service.Dependencies)
+
+	// Set NIMPipeline as the owner and controller of the NIMService
+	if err := controllerutil.SetControllerReference(nimPipeline, nimService, r.Scheme); err != nil {
+		return err
+	}
+
+	// Check if the NIMService already exists and create or update as necessary
+	existingService := &appsv1alpha1.NIMService{}
+	err := r.Get(ctx, types.NamespacedName{Name: nimService.Name, Namespace: nimService.Namespace}, existingService)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("Creating new NIMService", "name", nimService.Name)
+			return r.Create(ctx, nimService)
+		}
+		logger.Error(err, "Failed to get NIMService", "name", nimService.Name)
+		return err
+	}
+
+	// Update the existing NIMService's spec if needed
+	logger.Info("Updating existing NIMService", "name", nimService.Name)
+	existingService.Spec = nimService.Spec
```

Comment:
Should we compare to check if update is really needed?

Reply:
yes, good to check the hash before update. Fixed it.